### PR TITLE
fix the bounding rect of Arc

### PR DESCRIPTION
**Pull request by xubury**

### CODE REVIEW

**Comment by JamesBremner** [comment]
This is in reference to #15

**Comment by JamesBremner** [comment]
Please test your code using https://github.com/JamesBremner/DXF_Viewer/blob/master/dxf/arc.dxf as the input file.  It should look like this.

image

When I run this test with your code, the bounding rectangle does not work correctly.

I suspect that you are using a different interpretation of the start and end angles.  Unfortunately the DXF specs do not specify how these angles are to be understood ( https://documentation.help/AutoCAD-DXF/WS1a9193826455f5ff18cb41610ec0a2e719-7a35.htm ) - which is a common snag when working with DXF files.

In DXF-Viewer I have, somewhat arbitrarily, opted to use the interpretation documented here:

https://github.com/JamesBremner/DXF_Viewer/blob/ba0106c3fffba0444cebb3dd2175e7b5339cfe82/src/Arc.h#L10-L17

**Comment by JamesBremner** [comment]
My mistake.  There was an error in the arc drawing code.

**Comment by JamesBremner** [comment]
I have fixed the arc drawing error.  However, when I test your code it still does not work.  So I have added a simple test of the arc bounding box calculation.

https://github.com/JamesBremner/DXF_Viewer/blob/80b5c00e3fa11d04879d360dd35cce11e7e89919/build/codeblocks/UnitTest/main.cpp#L1-22

The output from this is

`50 50, 150 150`

This is clearly incorrect, since the arc segment is only in the top left and bottom left quadrants.

Again, I think you are using a different interpretation of the start and stop angles.  See post above for the meaning assigned to these angles by DXF viewer.

**Comment by xubury** [comment]
Oh, it was an error in my code. I completely forgot to deal with nagative angle which lead to a wrong quadrant calculation. However, I don't think there's any error in the arc drawing code, you might have misktaken the DXF file. I tested the dxf you provided and got something like this.
QQ截图20200311100612
It is odd cause it's not the arc you described. So I tried to create the dxf myself using a python module  `dxfwrite`.
```
drawing = dxf.drawing('drawing.dxf')
arc = dxf.arc(10.0, (5, 5), -45, 225)
drawing.add(arc)
drawing.save()
```
And as a result, I got this after I fixed the quadrant calculation.
QQ截图20200311100623

**Comment by xubury** [comment]
The new drawing arc code is conflicting with wxWidget draw function ` void wxDC::DrawEllipticArc(wxCoord x, wxCoord y, wxCoord w, wxCoord h, double sa, double ea)`. 
In wxWidgets doc, it says `x and y specify the x and y coordinates of the upper-left corner of the rectangle that contains the ellipse.` and `width and height specify the width and height of the rectangle that contains the ellipse.`
So the old code is correct for wxWidgets drawing. Perhaps there's an inconsistency between your different viewer.

**Comment by xubury** [comment]
There was no error in `Arc.cpp`(except for the bounding rect). But, there is an error in `windex.cpp`.
In `Arc.h`,  your comment says:
`draw.x1, draw.y1  Upper left corner of bounding rectangle of entire circle`
https://github.com/JamesBremner/DXF_Viewer/blob/80b5c00e3fa11d04879d360dd35cce11e7e89919/src/Arc.h#L42
And in `windex.cpp`, you have something like this:
https://github.com/JamesBremner/DXF_Viewer/blob/80b5c00e3fa11d04879d360dd35cce11e7e89919/src/windex.cpp#L170-L177
However, the function you defined in your windex project says the first and second argument should be the x and y for center. That's what make the arc drawing wrong.
```
    /** Draw Arc of circle
    @param[in] x for center, pixels 0 at left of window
    @param[in] y for center, pixels 0 at top of window
    @param[in] r radius, pixels
    @param[in] sa start angle degrees anti-clockwise from 3 o'clock
    @param[in] se end angle degrees anti-clockwise from 3 o'clock
    */
    void arc(
        int x, int y, double r,
        double sa, double ea )
```
 I think you should fix `windex.cpp` instead of `Arc.cpp`. This commit https://github.com/JamesBremner/DXF_Viewer/commit/566c5a55cc8b80c05256067c9ae1526e86d7c110 breaks the consistancy.

File: src/Arc.cpp [hunk 1]
```diff
@@ -31,8 +31,16 @@ CArc::~CArc()
 }
 void CArc::Update( cBoundingRectangle& rect )
 {
-    uint8_t start_quadrant = (int)sa % 360 / 90 + 1;// the quadrant of sa
-    uint8_t end_quadrant = (int)ea % 360 / 90 + 1;// the quadrant of ea
+    int8_t start_quadrant = (int)sa % 360 / 90 + 1;// the quadrant of sa
```
Comment (JamesBremner) [inline src/Arc.cpp]: You are doing rather a lot of casting between different integer types.  Why?  Surely, using int everywhere would make you code easier to understand and slightly faster?

Comment (xubury) [reply src/Arc.cpp]: yes, that make sense.

**Comment by JamesBremner** [comment]
You are using an incorrect ( i.e. different from that used by the rest of DVF-Viewer ) interpretation of the start and end angle, as shown by the following unit test code.  Please correct so I can merge your pull request.
```
    // test bounding box calculations for CArc

    dxfv::CArc A;
    A.x = 100;          // center at 100,100
    A.y = 100;
    A.r = 50;           // radius 50
    dxfv::cBoundingRectangle BR;

    std::cout << "bottom right quadrant\n";
    A.sa = 5;
    A.ea = 85;
    BR.init = false;
    A.Update( BR );
    std::cout << BR.x1<<" "<<BR.y1<<", "<<BR.x2<<" "<<BR.y2<<"\n";

    std::cout << "bottom left quadrant\n";
    A.sa = 95;
    A.ea = 175;
    BR.init = false;
    A.Update( BR );
    std::cout << BR.x1<<" "<<BR.y1<<", "<<BR.x2<<" "<<BR.y2<<"\n";

    std::cout << "top left quadrant\n";
    A.sa = 185;
    A.ea = 265;
    BR.init = false;
    A.Update( BR );
    std::cout << BR.x1<<" "<<BR.y1<<", "<<BR.x2<<" "<<BR.y2<<"\n";

    std::cout << "top right quadrant\n";
    A.sa = 375;
    A.ea = 355;
    BR.init = false;
    A.Update( BR );
    std::cout << BR.x1<<" "<<BR.y1<<", "<<BR.x2<<" "<<BR.y2<<"\n";
```

output is

```
bottom right quadrant
104.358 104.358, 149.81 149.81
bottom left quadrant
50.1903 104.358, 95.6422 149.81
top left quadrant
50.1903 50.1903, 95.6422 95.6422
top right quadrant
50 50, 149.81 150
```

**Pull request merged**